File: Cpp/2296_design_a_text_editor.cpp
// https://leetcode.cn/problems/design-a-text-editor/

#include <vector>
#include <string>

using namespace std;

class TextEditor {
public:
    TextEditor() {
        left_ = "";
        right_ = "";
    }
    
    void addText(string text) {
        left_ += text;
    }
    
    int deleteText(int k) {
        int n = left_.size();
        int res = min(k, n);
        left_.resize(n - res);
        return res;
    }
    
    string cursorLeft(int k) {
        while (k && left_.size()) {
            right_ += left_.back();
            left_.pop_back();
            k -= 1;
        }
        int start = max((int)left_.size()-10, 0);
        return left_.substr(start);
    }
    
    string cursorRight(int k) {
        while (k && right_.size()) {
            left_ += right_.back();
            right_.pop_back();
            k -= 1;
        }
        int start = max((int)left_.size()-10, 0);
        return left_.substr(start);
    }

    string left_, right_;
};

/**
 * Your TextEditor object will be instantiated and called as such:
 * TextEditor* obj = new TextEditor();
 * obj->addText(text);
 * int param_2 = obj->deleteText(k);
 * string param_3 = obj->cursorLeft(k);
 * string param_4 = obj->cursorRight(k);
 */

File: Python3/2296_design_a_text_editor.py
# https://leetcode.cn/problems/design-a-text-editor/

class TextEditor:
    def __init__(self) -> None:
        self.left = []
        self.right = []

    def addText(self, text: str) -> None:
        self.left.extend(text)
    
    def deleteText(self, k: int) -> int:
        pre = len(self.left)
        del self.left[-k:]
        return pre - len(self.left)
    
    def cursorLeft(self, k: int) -> str:
        while k and self.left:
            self.right.append(self.left.pop())
            k -= 1
        return "".join(self.left[-10:])
    
    def cursorRight(self, k: int) -> str:
        while k and self.right:
            self.left.append(self.right.pop())
            k -= 1
        return "".join(self.left[-10:])
    

# Your TextEditor object will be instantiated and called as such:
# obj = TextEditor()
# obj.addText(text)
# param_2 = obj.deleteText(k)
# param_3 = obj.cursorLeft(k)
# param_4 = obj.cursorRight(k)